### ISSUE TITLE
HAProxy 2.6 crashing when running on a Mikrotik container (ARM)

### ISSUE THREAD

**Issue by nazar554** [haproxy/haproxy#1984]
### Detailed Description of the Problem

I'm running HAProxy as a docker container on MicroTik HAP AC2 and it crashes after 3-5 hours of work.

### Expected Behavior

It runs without crashes.

### Steps to Reproduce the Behavior

1. Start the `haproxy:2.6.7-alpine` container
2. Wait for few hours
3. See it's stopped be cause of a crash

### What is your configuration?

```haproxy
global
	nbthread 4
	cpu-map auto:1/1-4 0-3
	log stdout format short daemon notice
	
	user haproxy
	group haproxy

	unix-bind prefix /var/lib/haproxy/ user haproxy group haproxy mode 660
	# Default SSL material locations
	ca-base ssl/certs
	crt-base ssl/private

	# generated 2020-12-22, Mozilla Guideline v5.6, HAProxy 2.2, OpenSSL 1.1.1d, intermediate configuration
	# https://ssl-config.mozilla.org/#server=haproxy&version=2.2&config=intermediate&openssl=1.1.1d&guideline=5.6
	ssl-default-bind-ciphers ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES256-GCM-SHA384:DHE-RSA-AES128-GCM-SHA256
	ssl-default-bind-ciphersuites TLS_AES_256_GCM_SHA384:TLS_AES_128_GCM_SHA256:TLS_CHACHA20_POLY1305_SHA256
	ssl-default-bind-options no-sslv3 no-tlsv10 no-tlsv11 no-tls-tickets

	ssl-default-server-ciphers ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES256-GCM-SHA384:DHE-RSA-AES128-GCM-SHA256
	ssl-default-server-ciphersuites TLS_AES_256_GCM_SHA384:TLS_AES_128_GCM_SHA256:TLS_CHACHA20_POLY1305_SHA256
	ssl-default-server-options no-sslv3 no-tlsv10 no-tlsv11 no-tls-tickets

	# curl https://ssl-config.mozilla.org/ffdhe2048.txt > /path/to/dhparam
	ssl-dh-param-file ssl/ffdhe4096.pem

defaults
	log global
	log stdout  format raw  local0  info

	mode	http
	option  srvtcpka
	option  clitcpka
	option	dontlognull
	option  splice-auto
	option  http-keep-alive
	option  redispatch
	option 	prefer-last-server
	retries                 3
	timeout http-request    10s
	timeout queue           2m
	timeout connect         10s
	timeout client          5m
	timeout server          5m
	timeout http-keep-alive 10s
	timeout check           10s

	errorfile 400 errors/400.http
	errorfile 403 errors/403.http
	errorfile 408 errors/408.http
	errorfile 500 errors/500.http
	errorfile 502 errors/502.http
	errorfile 503 errors/503.http
	errorfile 504 errors/504.http

frontend ft_https_tcp_sni
    bind [::]:443
    mode tcp

    tcp-request inspect-delay 5s
    tcp-request content capture req.ssl_sni len 253
    tcp-request content reject unless { req.ssl_hello_type 1 }
    tcp-request content set-var(req.host) req.ssl_sni,lower
    
    option tcplog

    acl is_sni_proxy var(req.host),map_str(ft_https_tcp_sni.map) -m found
    acl is_http_proxy var(req.host),map_str(ft_http.map) -m found

    use_backend %[var(req.host),map_str(ft_https_tcp_sni.map)] if is_sni_proxy
    use_backend bk_tcp_to_https if is_http_proxy
    default_backend bk_tcp_nohost

frontend ft_http
    mode http
    option  http-ignore-probes
    option  httplog

    bind [::]:80
    bind abns@haproxy_loopback_https.sock accept-proxy ssl strict-sni crt-list ft_https_crt.list
    capture request header Host len 253

    acl is_ssl_req ssl_fc
 
    http-request replace-header host ^(.*):80$ \1 if !is_ssl_req
    http-request replace-header host ^(.*):443$ \1 if is_ssl_req
 
    # prevent domain fronting
    http-request set-var(req.host) hdr(host)
    acl sni_host_matches ssl_fc_sni,strcmp(req.host) eq 0

    http-request reject if is_ssl_req !sni_host_matches
    
    http-request set-header Host %[var(req.host),lower]
    http-request set-header X-Forwarded-Proto http unless is_ssl_req
    http-request set-header X-Forwarded-Proto https if is_ssl_req
    http-request set-header X-Forwarded-Host %[hdr(host)]
    http-request set-header X-Forwarded-Proto-Version h2 if { ssl_fc_alpn -i h2 }
    http-request set-header X-Forwarded-Port %[dst_port]
    http-request set-header X-Forwarded-Path %[path]
    http-request set-header Forwarded for=\"[%[src]]\";host=%[hdr(host)];proto=%[hdr(X-Forwarded-Proto)]
    option forwardfor
    
    acl LOCALHOST6 src ::1/128

    acl is_haproxy hdr(host) -m str -i haproxy.example.com
    acl is_haproxy_bk_path pathq -m reg '^/\?stats([;&#]|$)'
    acl is_haproxy_bk_path path -m str /metrics
	
    http-request deny if is_haproxy !{ ssl_c_used } !LOCALHOST !LOCALHOST6

    use_backend bk_haproxy if is_haproxy is_haproxy_bk_path

    use_backend %[hdr(host),map_str(ft_http.map)]
    default_backend bk_http_nohost

backend bk_tcp_to_https
    mode tcp
    server sv_haproxy_loopback abns@haproxy_loopback_https.sock send-proxy-v2

backend bk_http_nohost
    mode http
    http-request set-log-level silent
    acl is_ssl_req ssl_fc
    http-request reject unless is_ssl_req
    
    # should be caught by bk_tcp_nohost
    http-response return status 503 default-errorfiles if is_ssl_req

backend bk_tcp_nohost
    mode tcp
    tcp-request content reject

backend bk_haproxy
    mode http
    acl is_ssl_cert_used ssl_c_used
   
    acl LOCALHOST6 src ::1/128     
    acl is_haproxy_metrics path -m str /metrics
     
    http-request deny unless LOCALHOST || LOCALHOST6 || is_ssl_cert_used
    http-request set-log-level silent if is_haproxy_metrics 
    http-request use-service prometheus-exporter if is_haproxy_metrics

    acl is_haproxy_stats pathq -m reg '^/\?stats([;&#]|$)'

    stats http-request deny unless is_ssl_cert_used
    stats enable
    stats admin if is_ssl_cert_used
    stats uri /?stats
    stats refresh 10s
    stats show-legends

    compression algo gzip
    compression type text/plain text/html text/css text/csv application/json application/javascript text/xml application/xml application/xml+rss text/javascript

backend bk_server1_http
  mode http
  acl is_ssl_req ssl_fc
  http-request redirect scheme https code 301 unless is_ssl_req
  http-request reject if is_ssl_req

backend bk_server1_https_tcp
  mode tcp
  server sv_server1 server1.example.com:443 check-sni server1.example.com ca-file ca.crt check inter 5s check-ssl verify required

backend bk_server2_http
  mode http
  acl is_ssl_req ssl_fc
  http-request redirect scheme https code 301 unless is_ssl_req
  http-request reject if is_ssl_req

backend bk_server2_https_tcp
  mode tcp
  server sv_server2 server2.example.com:443 check-sni server2.example.com ca-file ca.crt check inter 5s check-ssl verify required
```

### Output of `haproxy -vv`

```plain
HAProxy version 2.6.7-c55bfdb 2022/12/02 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2027.
Known bugs: http://www.haproxy.org/bugs/bugs-2.6.7.html
Running on: Linux 5.6.3 #2 SMP Mon Oct 17 11:19:47 UTC 2022 armv7l
Build options :
  TARGET  = linux-musl
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wundef -Wdeclaration-after-statement -Wfatal-errors -Wtype-limits -Wshift-negative-value -Wshift-ov
erflow=2 -Wduplicated-cond -Wnull-dereference -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-p
arameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wno-string-plus-int -Wno-atomic-alignment
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_PROMEX=1
  DEBUG   = -DDEBUG_STRICT -DDEBUG_MEMORY_POOLS

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL +THREAD -BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROX
Y +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H -ENGINE +GETADDRINFO +OPENSSL +LUA +ACCEPT4 -CLOSEFROM -ZLIB +SLZ +CPU_AFFINITY +TFO
 +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL -SYSTEMD -OBSOLETE_LINKER +PRCTL -PROCCTL +THREAD_DUMP -EVPORTS -OT -QUIC +PROMEX -MEMORY
_PROFILING

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=32, default=4).
Built with OpenSSL version : OpenSSL 3.0.7 1 Nov 2022
Running on OpenSSL version : OpenSSL 3.0.7 1 Nov 2022
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
OpenSSL providers loaded : default
Built with Lua version : Lua 5.3.6
Built with the Prometheus exporter as a service
Built with network namespace support.
Support for malloc_trim() is enabled.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.40 2022-04-14
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 12.2.1 20220924

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : prometheus-exporter
Available filters :
        [CACHE] cache
        [COMP] compression
        [FCGI] fcgi-app
        [SPOE] spoe
        [TRACE] trace
```

### Last Outputs and Backtraces

```plain
17:57:55 container,info,debug [NOTICE]   (1) : haproxy version is 2.6.7-c55bfdb
 17:57:56 container,info,debug [WARNING]  (1) : config : 'option prefer-last-server' ignored for backend 'bk_tcp_to_https' as it requ
ires HTTP mode.
 17:57:59 container,info,debug [WARNING]  (1) : config : 'option prefer-last-server' ignored for backend 'bk_tcp_nohost' as it requir
es HTTP mode.
 18:04:12 container,info,debug [WARNING]  (1) : config : 'option prefer-last-server' ignored for backend 'bk_server1_https_tcp' as it re
quires HTTP mode.
 18:10:34 container,info,debug [WARNING]  (1) : config : 'option prefer-last-server' ignored for backend 'bk_server2_https_tcp' as it r
equires HTTP mode.
 18:32:06 container,info,debug [NOTICE]   (1) : New worker (3) forked
 18:32:09 container,info,debug [NOTICE]   (1) : Loading success.
 18:34:01 container,info,debug [WARNING]  (3) : Server bk_server2_https_tcp/sv_server2 is DOWN, reason: Layer6 timeout, check duration: 1
1011ms. 0 active and 0 backup servers left. 0 sessions active, 0 requeued, 0 remaining in queue.
 18:34:04 container,info,debug <1>Server bk_server2_https_tcp/sv_server2 is DOWN, reason: Layer6 timeout, check duration: 11011ms. 0 acti
ve and 0 backup servers left. 0 sessions active, 0 requeued, 0 remaining in queue.
 18:34:04 container,info,debug Server bk_server2_https_tcp/sv_server2 is DOWN, reason: Layer6 timeout, check duration: 11011ms. 0 active 
and 0 backup servers left. 0 sessions active, 0 requeued, 0 remaining in queue.
 18:34:05 container,info,debug [ALERT]    (3) : backend 'bk_server2_https_tcp' has no server available!
 18:34:05 container,info,debug <0>backend bk_server2_https_tcp has no server available!
 18:34:05 container,info,debug backend bk_server2_https_tcp has no server available!
 18:34:29 container,info,debug [WARNING]  (3) : Server bk_server1_https_tcp/sv_server1 is DOWN, reason: Layer4 connection problem, info: "H
ost is unreachable", check duration: 11011ms. 0 active and 0 backup servers left. 0 sessions active, 0 requeued, 0 remaining in queue
.
 18:34:30 container,info,debug <1>Server bk_server1_https_tcp/sv_server1 is DOWN, reason: Layer4 connection problem, info: "Host is unreach
able", check duration: 11011ms. 0 active and 0 backup servers left. 0 sessions active, 0 requeued, 0 remaining in queue.
 18:34:30 container,info,debug Server bk_server1_https_tcp/sv_server1 is DOWN, reason: Layer4 connection problem, info: "Host is unreachabl
e", check duration: 11011ms. 0 active and 0 backup servers left. 0 sessions active, 0 requeued, 0 remaining in queue.
 18:34:37 container,info,debug [ALERT]    (3) : backend 'bk_server1_https_tcp' has no server available!
 18:34:41 container,info,debug <0>backend bk_server1_https_tcp has no server available!
 18:34:41 container,info,debug backend bk_server1_https_tcp has no server available!
 18:36:57 container,info,debug Thread 1 is about to kill the process.
 18:36:57 container,info,debug *>Thread 1 : id=0x76ff4020 act=1 glob=1 wq=0 rq=0 tl=0 tlsz=0 rqsz=0
 18:36:57 container,info,debug       1/1    stuck=1 prof=0 harmless=0 wantrdv=0
 18:36:57 container,info,debug              cpu_ns: poll=6079841828 now=8615461951 diff=2535620123
 18:36:57 container,info,debug              curr_task=0x76429f10 (tasklet) calls=1
 18:36:57 container,info,debug                fct=0x43fee0(ssl_sock_io_cb) ctx=0x76932060
 18:36:57 container,info,debug 
 18:36:57 container,info,debug   Thread 2 : id=0x763c2880 act=1 glob=1 wq=0 rq=0 tl=0 tlsz=0 rqsz=0
 18:36:57 container,info,debug       1/2    stuck=0 prof=0 harmless=0 wantrdv=0
 18:36:57 container,info,debug              cpu_ns: poll=2639222250 now=5349251254 diff=2710029004
 18:36:57 container,info,debug              curr_task=0
 18:36:57 container,info,debug   Thread 3 : id=0x7639d880 act=1 glob=1 wq=0 rq=0 tl=0 tlsz=0 rqsz=0
 18:36:57 container,info,debug       1/3    stuck=0 prof=0 harmless=0 wantrdv=0
 18:36:57 container,info,debug              cpu_ns: poll=4172504345 now=7753570838 diff=3581066493
 18:36:57 container,info,debug              curr_task=0
 18:36:57 container,info,debug   Thread 4 : id=0x76378880 act=1 glob=1 wq=0 rq=0 tl=0 tlsz=0 rqsz=0
 18:36:57 container,info,debug       1/4    stuck=0 prof=0 harmless=0 wantrdv=0
 18:36:57 container,info,debug              cpu_ns: poll=9497090643 now=15785078925 diff=6287988282
 18:36:57 container,info,debug              curr_task=0x768bef20 (task) calls=8 last=0
 18:36:57 container,info,debug                fct=0x5a4868(process_chk) ctx=0x76905430
 18:36:57 container,info,debug 
 19:01:38 container,info,debug [NOTICE]   (1) : haproxy version is 2.6.7-c55bfdb
 19:01:44 container,info,debug [ALERT]    (1) : Current worker (3) exited with code 134 (Aborted)
 19:01:48 container,info,debug [ALERT]    (1) : exit-on-failure: killing every processes with SIGTERM
 19:01:50 container,info,debug [WARNING]  (1) : All workers exited. Exiting... (134)
```

### Additional Information

uname -a
`Linux haproxy 5.6.3 #2 SMP Mon Oct 17 11:19:47 UTC 2022 armv7l Linux`

**Comment by lukastribus** [comment]
Do you mean the `RBD52G-5HacD2HnD-TC` with 128MB RAM? I don't think that's enough memory for RouterOS7 + an alpine container.

**Comment by TimWolla** [comment]
> 18:36:57 container,info,debug Thread 1 is about to kill the process.

And to clarify: HAProxy dies because the watchdog triggered, because a thread was blocked for too long. It looks like the thread is blocked during TLS operations. Based on Lukas' reply it might be that the CPU might not be sufficiently fast for TLS with your traffic levels.

**Comment by nazar554** [comment]
I just recently upgraded to RouterOS 7, trying to get my haproxy setup working (I used a Orange Pi Zero before).

I'm running with `-m 64` and I didn't get any OOM crashes yet. Here is top output inside container.

```
Mem total:119372 anon:16584 map:18348 free:7344
 slab:27060 buf:10956 cache:45564 dirty:16 write:0
Swap total:0 free:0
  PID   VSZ^VSZRW^  RSS (SHR) DIRTY (SHR) STACK COMMAND
    3 13656  5920  7920  4488  2804     0   132 haproxy -W -db -f /usr/local/etc/haproxy/haproxy.cfg -m 64
    1 11040  3328  6964  4492  1744     0   132 haproxy -W -db -f /usr/local/etc/haproxy/haproxy.cfg -m 64
   13  1784   200  1160   988   108     0   132 /bin/sh -i
   18  1752   168  1036   988    48     0   132 top
```

Hmm the TLS is there only for server checks and `301 Redirects`, so it's timing out during a handshake?
I guess I could try switching to TCP-only checks.

Also is the `tcp-request content capture req.ssl_sni` stuff also CPU intensive? Should it be okay to run it as a SNI proxy?

**Comment by lukastribus** [comment]
If you run with `-m 64`, you probably would never see any OOM's for this exact reason: haproxy won't allocate more than that, but I assume this comes at a cost (in CPU cycles or capacity):

    -m <limit> : limit the total allocatable memory to <limit> megabytes across
        all processes. This may cause some connection refusals or some slowdowns
        depending on the amount of memory needed for normal operations. 

I would:

- set appropriate (low) maxconn values (global, per frontend and per backend server)
- disable compression
- downgrade to `nbthread 1`

Compression and TLS termination are heavy CPU users (and we are using OpenSSL 3.0 here, which makes things worse in multi-threaded environments).

None of this is to say that we can definitely exclude a haproxy bug however.


Could give latest 2.7 a try also? Some important improvements have gone into it like #1902


> Also is the `tcp-request content capture req.ssl_sni` stuff also CPU intensive? Should it be okay to run it as a SNI proxy?

No, it shouldn't be particularly CPU intensive, no.

**Comment by wtarreau** [comment]
While I don't find that such a machine with 4 cores at 700 MHz is that small, and I'm routinely using it on single-core MIPS at 396 MHz, it's true that we can't rule out a long locking issue with openssl 3, and the crashing function (`ssl_io_cb`) tends to support this! Sadly on musl we don't have `backtrace()` so we don't know exactly where the crash happened, but it might be worth trying our own in-house implementation by building with `USE_BACKTRACE=1`. It may work or may fail. But it would tell us way more about where the issue happened if it works. For example on a minimal libc another issue can be the memory allocator. OpenSSL makes extreme use of malloc() (even for a few bytes) and it could be possible that it periodically needs to defragment itself and takes more time at this for example, or is simply stuck in contention between several threads.

As Lukas suggested, trying with nbthread 1 first would be useful.

It's also perfectly possible that we have a bug that is revealed on this machine, but for using a lot of ARM myself, I'm relatively confident that we would probably already have noticed it.

If you're willing to experiment, in 2.7 we've also reimplement our own rwlocks to replace the pthread ones that OpenSSL abuses. Normally they're provided by libpthread that comes with the libc. It's possible that a minimal and/or sub-optimal implementation is provided with musl that collapses under heavy use. In that case, building with USE_PTHREAD_EMULATION=1 would replace it with ours so you can compare.

Also, please have a look at your kernel messages with `dmesg`, in case any report of OOM or any other event we're not thinking of would ring a bell.

**Comment by wtarreau** [comment]
Another point, such small embedded systems sometimes come with background tasks to download updates and whatever. At least just for this it would be wise to have less haproxy threads than there are physical cores in the CPU (4), so that you don't end up with long pauses of a thread that holds a lock and blocks another one. Unless you need extreme SSL performance (which I doubt on such a small system), I'd suggest to go no higher than 2 threads anyway (but only once the problem is identified).

**Comment by NicolasCARPi** [comment]
Hello,

I believe I am facing a very similar issue, still present with 2.7.1.

Context: running on x86 with 8 Gb of RAM and 4 CPU, Rocky Linux 9, so more resources than what @nazar554 has. HAProxy is the Docker/Alpine version.

Still, I can reliably trigger a crash when I try to load netdata (a web service), which is consuming a lot of CPU (see [this post on netdata forum](https://community.netdata.cloud/t/current-issues-under-rocky-9-and-docker/3667/14?u=smap)).

The interesting thing is that setting `nbthread 1` will prevent the crash. It still causes netdata to behave wrongly, but the interesting thing for you is that there is a reliably triggered crash when we have more than one thread.

The logs of the crash:

~~~
/0 0/0 "GET /netdata/api/v1/data?chart=system.swap&_=1673492118754&format=array&points=240&group=average&gtime=0&options=absolute%7Cpercentage%7Cjsonwrap%7Cnonzero&after=-240&dimensions=used HTTP/1.1"
Thread 2 is about to kill the process.
  Thread 1 : id=0x7feb856ca9d0 act=0 glob=0 wq=1 rq=0 tl=0 tlsz=0 rqsz=0
      1/1    stuck=0 prof=0 harmless=1 wantrdv=0
             cpu_ns: poll=883220847 now=904301108 diff=21080261
             curr_task=0
* Thread 2 : id=0x7feb848a0b30 act=1 glob=1 wq=1 rq=0 tl=0 tlsz=0 rqsz=2
      1/2    stuck=0 prof=0 harmless=0 wantrdv=0
             cpu_ns: poll=9857209 now=3002220932 diff=2992363723
             curr_task=0x7feb85565cd0 (task) calls=1 last=0
               fct=0x556e6c2b02b0(process_stream) ctx=0x7feb848db770
             strm=0x7feb848db770,c08 src=192.168.1.76 fe=http-in be=netdata dst=unknown
             txn=0x7feb849012e0,3000 txn.req=MSG_BODY,4c txn.rsp=MSG_RPBEFORE,0
             rqf=64d08002 rqa=60900 rpf=a0000000 rpa=20000000
             scf=0x7feb84b388d0,EST,20 scb=0x7feb848eccc0,INI,21
             af=0,0 sab=0,0
             cof=0x7feb85556b50,80000300:H1(0x7feb855566a0)/RAW(0)/tcpv4(20)
             cob=0,0:NONE(0)/NONE(0)/NONE(-1)

  Thread 3 : id=0x7feb8487bb30 act=0 glob=0 wq=1 rq=0 tl=0 tlsz=0 rqsz=0
      1/3    stuck=0 prof=0 harmless=1 wantrdv=0
             cpu_ns: poll=2571520629 now=2589997468 diff=18476839
             curr_task=0
  Thread 4 : id=0x7feb84856b30 act=0 glob=0 wq=1 rq=0 tl=0 tlsz=0 rqsz=0
      1/4    stuck=0 prof=0 harmless=1 wantrdv=0
             cpu_ns: poll=1785718262 now=1797374678 diff=11656416
             curr_task=0
[NOTICE]   (1) : haproxy version is 2.7.1-3e4af0e
[ALERT]    (1) : Current worker (8) exited with code 134 (Aborted)
[ALERT]    (1) : exit-on-failure: killing every processes with SIGTERM
[WARNING]  (1) : All workers exited. Exiting... (134)
~~~

I have no idea if the above can give useful information on the cause of the crash.


Haproxy -vv

~~~
HAProxy version 2.7.1-3e4af0e 2022/12/19 - https://haproxy.org/
Status: stable branch - will stop receiving fixes around Q1 2024.
Known bugs: http://www.haproxy.org/bugs/bugs-2.7.1.html
Running on: Linux 5.14.0-162.6.1.el9_1.0.1.x86_64 #1 SMP PREEMPT_DYNAMIC Mon Nov 28 18:44:09 UTC 2022 x86_64
Build options :
  TARGET  = linux-musl
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wundef -Wdeclaration-after-statement -Wfatal-errors -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wno-string-plus-int -Wno-atomic-alignment
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_PROMEX=1
  DEBUG   = -DDEBUG_STRICT -DDEBUG_MEMORY_POOLS

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL +THREAD -PTHREAD_EMULATION -BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H -ENGINE +GETADDRINFO +OPENSSL -OPENSSL_WOLFSSL +LUA +ACCEPT4 -CLOSEFROM -ZLIB +SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL -SYSTEMD -OBSOLETE_LINKER +PRCTL -PROCCTL +THREAD_DUMP -EVPORTS -OT -QUIC +PROMEX -MEMORY_PROFILING +SHM_OPEN

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_TGROUPS=16, MAX_THREADS=256, default=4).
Built with OpenSSL version : OpenSSL 3.0.7 1 Nov 2022
Running on OpenSSL version : OpenSSL 3.0.7 1 Nov 2022
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
OpenSSL providers loaded : default
Built with Lua version : Lua 5.3.6
Built with the Prometheus exporter as a service
Built with network namespace support.
Support for malloc_trim() is enabled.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.42 2022-12-11
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 12.2.1 20220924

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : prometheus-exporter
Available filters :
        [BWLIM] bwlim-in
        [BWLIM] bwlim-out
        [CACHE] cache
        [COMP] compression
        [FCGI] fcgi-app
        [SPOE] spoe
        [TRACE] trace
~~~

Please let me know if I can be of further help by providing more debug stuff. I might try later to setup a reproducible server (that is not prod!), with Rocky 9, Docker haproxy alpine and netdata, so that I can probably more easily give dumps and trace and whatnot. But maybe you already have an idea of what to try.

I tried reducing maxconn from 5000 to 2000, or removing the "check port" in the server line in backend, but no impact. I don't believe that I have a particularly stringent linux configuration, and I see nothing in dmesg. Given that with 1 thread the process lives happily ever after, the possibility of a bug in HAProxy is non null.

Best,
~Nico

**Comment by wtarreau** [comment]
Hi Nico,

do you have the core file, to open under gdb and get the full backtrace (using `t a a bt full`) ?
Such a crash definitely is a bug, there's no doubt about this, but what could trigger it is the question! I don't know "netdata". I don't know what it's doing specific that could trigger this problem either, so I'm a bit lost not even having any idea what to look for at this point. And we're having an issue with crash dumps in 2.7 not showing the backtrace by default (fixed in 2.8-dev, will be backported). Regardless, the backtrace in gdb would indicate line numbers and maybe even some variables we'll depend on.

**Comment by nazar554** [comment]
It's possible that RouterOS applied `ram-high` limit throttles the cgroup and it crashes during SSL handshake.
I'm not seeing the bug once I removed that option.
I'll try running different combinations of 2.6.7, 2.7.1 and 2.8-dev1 (both alpine/glibc) to see if I can reproduce the crash again

**Comment by NicolasCARPi** [comment]
> do you have the core file

No, and I don't know how to get it. I tried adding `set-dumpable` in the config, also removed the `read_only=true` for the container, but the errors in the log don't mention something like "core was dumped to...". What is the default path it would be written to? The container exits and is restarted by docker, so I would probably need to mount this path in the host. I also cannot really attach to the process because of the docker thing (not root, no gdb, etc...). Unfortunately, this bug is not reproducible locally on my computer, while it is 100% reproducible on my servers, and very easy to trigger, so it would be great if I was able to produce meaningful trace from the servers.

> I don't know what it's doing specific that could trigger this problem either,

Generating a bunch of graphs by collecting many data points from various sources. And also sending stuff to the could endpoint.

**Comment by TimWolla** [comment]
@NicolasCARPi Is Netdata running on the same host as HAProxy?

**Comment by NicolasCARPi** [comment]
@TimWolla yes!

**Comment by wtarreau** [comment]
The core should be dumped to a location defined in `/proc/sys/kernel/core_pattern`, it's very possible that on your distro this starts with a pipe followed by a command corresponding to a tool that is not even installed, as that's common :-/  Either you'll need to install the associated tool (which may even help retrieve it out of the container), or you'll have to write a regular file pattern written in a location with write permissions for the dumping (e.g. `/var/core/core.%p` after `chmod 1777 /var/core`).

**Comment by NicolasCARPi** [comment]
So the `core_pattern` in the container is `|/usr/lib/systemd/systemd-coredump %P %u %g %s %t %c %h` and as it turns out, the dumps correctly appear in the host in `/var/lib/systemd/coredump`. Thanks to the shared kernel I guess.

Unfortunately, I don't think they provide anything useful because they're coming from an alien system (the container), so gdb has a hard time figuring things out.

Here is a snippet of what I get:

~~~log
                Stack trace of thread 8:
                #0  0x00007f79d22bc22d n/a (/lib/ld-musl-x86_64.so.1 + 0x4922d)
                #1  0x0000000097661c90 n/a (n/a + 0x0)
                ELF object binary architecture: AMD x86-64
~~~

I'll continue to explore different options to try and get something useful. Is there some haproxy build options that can help get full debug, or something that comes to mind?

I think it would be super helpful to have a specific docker tag "debug" that can be readily used by users to debug stuff more easily. This image would contain all debug information/tools directly.

I'm attaching the dump as a zip file here anyway. [core.zip](https://github.com/haproxy/haproxy/files/10425544/core.zip)

Maybe someone can figure things out from this.

**Comment by NicolasCARPi** [comment]
So the `core_pattern` in the container is `|/usr/lib/systemd/systemd-coredump %P %u %g %s %t %c %h` and as it turns out, the dumps correctly appear in the host in `/var/lib/systemd/coredump`. Thanks to the shared kernel I guess.

Unfortunately, I don't think they provide anything useful because they're coming from an alien system (the container), so gdb has a hard time figuring things out.

Here is a snippet of what I get:

~~~log
                Stack trace of thread 8:
                #0  0x00007f79d22bc22d n/a (/lib/ld-musl-x86_64.so.1 + 0x4922d)
                #1  0x0000000097661c90 n/a (n/a + 0x0)
                ELF object binary architecture: AMD x86-64
~~~

I'll continue to explore different options to try and get something useful. Is there some haproxy build options that can help get full debug, or something that comes to mind?

I think it would be super helpful to have a specific docker tag "debug" that can be readily used by users to debug stuff more easily. This image would contain all debug information/tools directly.

**Comment by wtarreau** [comment]
Nicolas,

could you please do me a favor and run `objdump --disassemble=thread_isolate haproxy` on the haproxy executable you're using ? I'm seeing that you're using libmusl, I don't know if sched_yield() is being used there, and am interested in looking at how the loop was optimized, and particularly if .threads_enabled[] is being reloaded on each round or not. What I'm suspecting is that one thread is gone but others didn't see it leave while they were ready to do something which required synchronization.

**Comment by NicolasCARPi** [comment]
Willy,

Here it is:

~~~
objdump --disassemble=thread_isolate haproxy

haproxy:     file format elf64-x86-64


Disassembly of section .init:

Disassembly of section .plt:

Disassembly of section .plt.got:

Disassembly of section .text:

00000000001fc6a0 <thread_isolate>:
  1fc6a0:       64 48 8b 04 25 68 d9    mov    %fs:0xffffffffffffd968,%rax
  1fc6a7:       ff ff 
  1fc6a9:       f0 48 09 05 df 8f 19    lock or %rax,0x198fdf(%rip)        # 395690 <threads_harmless_mask>
  1fc6b0:       00 
  1fc6b1:       64 48 8b 04 25 68 d9    mov    %fs:0xffffffffffffd968,%rax
  1fc6b8:       ff ff 
  1fc6ba:       f0 48 09 05 2e 9a 0e    lock or %rax,0xe9a2e(%rip)        # 2e60f0 <threads_want_rdv_mask>
  1fc6c1:       00 
  1fc6c2:       48 8b 05 c7 8f 19 00    mov    0x198fc7(%rip),%rax        # 395690 <threads_harmless_mask>
  1fc6c9:       48 8b 15 18 9a 0e 00    mov    0xe9a18(%rip),%rdx        # 2e60e8 <all_threads_mask>
  1fc6d0:       48 8b 0d 11 9a 0e 00    mov    0xe9a11(%rip),%rcx        # 2e60e8 <all_threads_mask>
  1fc6d7:       48 21 c2                and    %rax,%rdx
  1fc6da:       48 39 ca                cmp    %rcx,%rdx
  1fc6dd:       75 21                   jne    1fc700 <thread_isolate+0x60>
  1fc6df:       64 48 8b 14 25 68 d9    mov    %fs:0xffffffffffffd968,%rdx
  1fc6e6:       ff ff 
  1fc6e8:       48 f7 d2                not    %rdx
  1fc6eb:       48 21 c2                and    %rax,%rdx
  1fc6ee:       f0 48 0f b1 15 99 8f    lock cmpxchg %rdx,0x198f99(%rip)        # 395690 <threads_harmless_mask>
  1fc6f5:       19 00 
  1fc6f7:       75 0e                   jne    1fc707 <thread_isolate+0x67>
  1fc6f9:       c3                      retq   
  1fc6fa:       66 0f 1f 44 00 00       nopw   0x0(%rax,%rax,1)
  1fc700:       48 8b 05 89 8f 19 00    mov    0x198f89(%rip),%rax        # 395690 <threads_harmless_mask>
  1fc707:       f3 90                   pause  
  1fc709:       eb be                   jmp    1fc6c9 <thread_isolate+0x29>

Disassembly of section .fini:
~~~

**Comment by wtarreau** [comment]
Perfect, thank you, so there's no sched_yield(), and despite this values are properly reloaded, that was my concern.

**Comment by wtarreau** [comment]
Nicolas, I found some bugs that could cause an old process to sometimes loop on reloads in 2.7. I've merged the fix in 2.7-maint that I hope to finally tag as 2.7.2 tomorrow. I don't know if it's related to what you're seeing since you don't seem to have reloads involved there, but it should not be exclusive to reloads and could theoretically also happen in other situations dealing with thread isolation and panics. Seeing the number of differences between 2.6 and 2.7 makes me realize that it's unlikely that your issue is the same as the orginal reporter's, and for us it becomes complex to consider the input conditions since they differ between the two reports. Thus what I'm proposing is that you retry on 2.7.2 once it's released, and if the issue still happens, then you'd create a separate issue. Thanks!

**Comment by NicolasCARPi** [comment]
Willy, ok sure, I'll try 2.7.2 and open a fresh issue or report here if it fixed it!

**Comment by capflam** [comment]
Any news about this issue ? Is it still relevant ?

**Comment by nazar554** [comment]
> Any news about this issue ? Is it still relevant ?

Doesn't reproduce after I removed `ram-high` cgroup limit on RouterOS
@NicolasCARPi might have a different repro approach

**Comment by NicolasCARPi** [comment]
My issue was not the same, see #2008.

**Comment by capflam** [comment]
Thanks guys ! So I guess we can close this issue.
